src/nodes.coffee

```diff
@@ -1894,6 +1897,66 @@ exports.Obj = class Obj extends Base
       answer.push @makeCode join
     if @front then @wrapInParentheses answer else answer
 
+  # Convert "bare" properties to `ObjectProperty`s (or `Splat`s)
+  expandProperty: (property) ->
+    {variable, context, operatorToken} = property
+    key = if property instanceof Assign and context is 'object'
+      variable
+    else if property instanceof Assign
+      operatorToken.error "unexpected #{operatorToken.value}" unless @lhs
+      variable
+    else
+      property
+    if key instanceof Value and key.hasProperties()
+      key.error 'invalid object key' unless context isnt 'object' and key.this
+      if property instanceof Assign
+        return new ObjectProperty fromAssign: property
+      else
+        return new ObjectProperty key: property
+    return new ObjectProperty(fromAssign: property) unless key is property
+    return property if property instanceof Splat
+
+    new ObjectProperty key: property
+
+  expandProperties: ->
+    @expandProperty(property) for property in @properties
+
+  astType: -> 'ObjectExpression'
+
+  astProperties: ->
+    properties:
+      property.ast() for property in @expandProperties()
+    implicit: !!@generated
+
+exports.ObjectProperty = class ObjectProperty extends Base
+  constructor: ({key, fromAssign}) ->
```

Comment:
`ObjectProperty::constructor()` is a little more "clever" than a typical node class constructor, knowing how to deconstruct an `Assign` and self-initializing its own `@locationData`

The target shape is basically:
- `@key` is always initialized
- `@value` is initialized for everything except "simple" shorthand properties (eg `{a, [b], @c}`)
- `@shorthand` is set to `true` for shorthand properties (ie when `shorthand: true` in the AST)
- `@locationData` is initialized

This makes the conversion to AST (ie `ObjectProperty::astProperties()`) straightforward